### PR TITLE
Hide internal details of ActionListener impls

### PR DESCRIPTION
**Pull request by DaveCTurner**
The `ActionListener` interface implements its static methods using various inner classes. These inner classes are public because interfaces do not support package-private inner classes, although they are `final` and only have `protected` constructors which prevent anyone else from instantiating them. It's certainly a mistake for code in other packages to access these things.

This commit moves the implementation details into a separate class to hide them from the rest of the world. It also tightens up the calls to `onFailure()` in a few places to better handle the possibility of this method throwing an exception. Finally, it moves `DelegatingActionListener` to a top level public class since this utility is used in quite a number of places.

### CODE REVIEW

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java [hunk 1]
```diff
@@ -1511,6 +1511,7 @@ public void testRealmAuthenticateTerminateAuthenticationProcessWithException() {
         verifyNoMoreInteractions(operatorPrivilegesService);
     }
 
+    @AwaitsFix(bugUrl = "TODO")
```
Comment (DaveCTurner) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java:1514]: We deliberately call `onFailure(null)` here:

https://github.com/elastic/elasticsearch/blob/bb44f7f30392f4a4f9ae418757092aa653898e62/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RealmsAuthenticator.java#L186

I am suppressing this test temporarily to see if there are any other places that use a null exception, but will revert this change before marking this as ready-to-review.

Comment (DaveCTurner) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java]: I opened https://github.com/elastic/elasticsearch/pull/94328 to address this; apart from that this is ready to review

Comment (DaveCTurner) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java]: https://github.com/elastic/elasticsearch/pull/94328 is merged so I've reinstated the null check

File: server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java [hunk 2]
```diff
@@ -19,21 +19,22 @@
 /**
  * An action listener that wraps another action listener and dispatches its completion to an executor.
  */
-public final class ThreadedActionListener<Response> extends ActionListener.Delegating<Response, Response> {
+public final class ThreadedActionListener<Response> implements ActionListener<Response> {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java:22]: Just curious: why does this need to change? Isn't the implementation still delegating?

**Review by rjernst — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java:22]: We're not using anything from `DelegatingActionListener` here - we completely override both `onFailure` and `toString`.

**Pull request merged**